### PR TITLE
vc: Remove device when AddDevice encounters error

### LINKED ISSUE #2295: Remove device when AddDevice encounters error
# Description of problem

If Sandbox#AddDevice encounters error, the error is returned. However, the new device is left behind.

# Expected result

Device should be detached / removed when error is encountered.

# Actual result

### PR DESCRIPTION
**Pull request by tedyu**
Fixes #2295

Signed-off-by: Ted Yu <yuzhihong@gmail.com>

### CODE REVIEW

File: virtcontainers/sandbox.go [hunk 1]
```diff
@@ -1898,11 +1898,31 @@ func (s *Sandbox) AddDevice(info config.DeviceInfo) (api.Device, error) {
...
+					"sandbox":   s.id,
+					"device-ID": b.DeviceID(),
+			    	}).Error("failed to remove device")
+		}
 		return nil, err
 	}
 
 	if !s.supportNewStore() {
 		if err := s.storeSandboxDevices(); err != nil {
+			if err1 := s.devManager.DetachDevice(b.DeviceID(), s); err1 != nil {
```
Comment (bergwolf) [inline virtcontainers/sandbox.go]: Could you put these in two defer functions?

Comment (tedyu) [reply virtcontainers/sandbox.go]: If DetachDevice returns error, I don't think RemoveDevice should be called.

Can you elaborate on how the defer functions should be formed ?

I think the current form is more readable.

Comment (bergwolf) [reply virtcontainers/sandbox.go]: NewDevice()
defer RemoveDevice()
AttachDevice()
defer DetachDevice()

Comment (tedyu) [reply virtcontainers/sandbox.go]: It seems two error variables should be used if we go along the above route.

If DetachDevice call fails, should RemoveDevice be called ?

I think continuing with RemoveDevice call would complicate troubleshooting.

**Comment by tedyu** [comment]
@jodh-intel @grahamwhaley 
Please review.

**Comment by tedyu** [comment]
@bergwolf 
Your comment has been addressed.

**Comment by tedyu** [comment]
@amshinde @jodh-intel 
Can you help trigger the ubuntu test ?

thanks

**Comment by tedyu** [comment]
For the centos run:
```
W1204 09:18:35.984440   79929 removeetcdmember.go:61] [reset] failed to remove etcd member: error syncing endpoints with etc: etcdclient: no available endpoints
```
I don't think the above was related to my change.

@bergwolf 
Can you trigger another test run ?

Thanks

**Comment by tedyu** [comment]
@bergwolf 
jenkins-ci-ARM-ubuntu-18-04 seemed to have similar error.
Please trigger test run for that as well.

Thanks

**Pull request merged**